cmd/axelard/cmd/vald/broadcaster/broadcast.go

```diff
@@ -17,60 +20,170 @@ import (
+		var batch []broadcastTask
+		msgCount := 0
+		for {
+			// we cannot split a single task, so take at least on task and then fill up the batch
+			// until the size limit is reached or no new tasks are in the backlog
+			batchWouldBeTooLarge := len(batch) > 0 && msgCount+len(b.backlog.Peek().Msgs) > b.batchSizeLimit
+			if batchWouldBeTooLarge || b.backlog.Len() == 0 {
+				break
+			}
+
+			task := b.backlog.Pop()
+
+			batch = append(batch, task)
+			msgCount += len(task.Msgs)
 		}
 
-		b.logger.Debug(fmt.Sprintf("tx response with hash [%s] and opcode [%d]: %s",
-			response.TxHash, response.Code, response.RawLog))
+		b.logger.Debug("high traffic; merging batches", "batch_size", msgCount)
+		b.broadcastWithRetry(batch...)
+	}
+}
+
+// Broadcast queues up the given messages for broadcast to the network. This function in thread-safe, blocks until it gets a response.
+func (b *Broadcaster) Broadcast(ctx context.Context, msgs ...sdk.Msg) (*sdk.TxResponse, error) {
+	if len(msgs) == 0 {
+		return nil, fmt.Errorf("no messages to broadcast")
+	}
 
-		// broadcast has been successful, so increment sequence number
-		b.txFactory = txf.WithSequence(txf.Sequence() + 1)
+	// serialize concurrent calls to broadcast
+	callback := make(chan broadcastResult, 1)
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-b.backlog.Push(broadcastTask{ctx, msgs, callback}):
+		b.logger.Debug("queuing up messages", "msg_count", len(msgs))
+		break
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res := <-callback:
+		return res.Response, res.Err
+	}
+}
 
-		return nil
-	}, func(err error) bool {
-		if !utils.IsABCIError(err) {
-			return true
+func (b *Broadcaster) broadcastWithRetry(tasks ...broadcastTask) {
+	var response *sdk.TxResponse
+
+	var msgs []sdk.Msg
+	for _, task := range tasks {
+		if task.Ctx.Err() != nil {
+			continue
 		}
+		msgs = append(msgs, task.Msgs...)
+	}
+
+	err := b.retryPipeline.Push(
+		func() error {
+			logger := b.logger.With("batch_size", len(msgs))
+			logger.Debug("starting to broadcast message batch")
+			txf, err := prepareFactory(b.clientCtx, b.txFactory)
+			if err != nil {
+				return err
+			}
+
+			response, err = Broadcast(b.clientCtx, txf, msgs)
+			if err != nil {
+				// reset account and sequence number in case they were the issue
+				b.txFactory = b.txFactory.
+					WithAccountNumber(0).
+					WithSequence(0)
+				return err
+			}
+
+			logger.Debug("received tx response",
+				"hash", response.TxHash,
+				"op_code", response.Code,
+				"raw_log", response.RawLog,
+			)
+
+			// broadcast has been successful, so increment sequence number
+			b.txFactory = txf.WithSequence(txf.Sequence() + 1)
+
+			return nil
+		},
+		func(err error) bool {
+			logger := b.logger.With("batch_size", len(msgs))
+			i, ok := tryParseErrorMsgIndex(err)
+			if ok && len(msgs) > 1 {
+				logger.Debug(fmt.Sprintf("excluding message at index %d due to error", i))
+				msgs = append(msgs[:i], msgs[i+1:]...)
```

Comment:
Where is the failed msg put back into the queue for a retry?

Reply:
the retry pipeline does that automatically